044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,253 @@
+smarter about only throttling writes to the brokers running out of space, so we fence the whole cluster.
+
+### High level changes
+To better support external sources for managing quotas this proposal introduces some new concepts to the plugin:
+
+#### Quota Source
+
+We propose adding a Quota Source concept to the plugin to provide an extension point where we could plug in
+future sources of quotas, like pull them from an external system.
+
+#### Throttle Factor
+
+We introduce the concept of throttle factor. This is a factor in the range (0.0, 1.0) that is applied to a quota to
+calculate the final byte rate limit. 
+
+This is a concept that was already implicit in the current calculations, but we want to name it so that we can use it in 
+metric names and provide an extension point in case we want to externalise the Throttle Factor.
+
+#### Volume
+
+The storage quotas operate on observations about a **Volume** with these characteristics:
+1. `logDir`: the path of the logdir
+2. `brokerId`
+3. `capacityBytes`
+4. `usableBytes`
+
+#### Volume Source
+
+The quota plugin operates on observations about Volumes, the **Volume Source** determines where
+we obtain those observations. Values are:
+1. `local`: we inspect this brokers local log dirs to discover the state of their Volume
+2. `cluster`: we ask kafka for the state of all log dirs for all currently active nodes
+
+the source will be configurable with a property like:
+
+1. `client.quota.callback.static.storage.volume.source=cluster`
+
+Local will be the default if the property isn't provided but is deprecated from the beginning.
+
+#### Cluster Volume Source
+
+With the introduction of [KIP-827](https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+logdirs+total+and+usable+space+via+Kafka+API) in
+kafka 3.3 we can now obtain the total and usable bytes per log dir as part of the DescribeLogDirsResponse.
+
+Note: if a single disk contains multiple log dirs, it will be described multiple times through the kafka APIs. This
+repetition is acceptable as our new limit types will be applied per-volume, so redundant volume descriptions don't
+impact the outcome.
+
+The Cluster Volume Source will use this API to discover volume information for the whole cluster.
+
+##### Admin Client Configuration
+
+To obtain log dir descriptions though the admin api we need to construct an admin client. We intend to
+connect to the replication listener and will piggyback on the broker's existing SSL configuration to
+obtain keystore/truststore properties.
+
+If using cluster sourced volumes we will expose these optional properties to customise the admin connection:
+
+1. `client.quota.callback.kafka.listener.name` (default `replication-9091`)
+2. `client.quota.callback.kafka.listener.port` (default `9091`)
+3. `client.quota.callback.kafka.listener.protocol` (default `SSL`)
+4. `client.quota.callback.kafka.clientIdPrefix` (default `__strimzi`)
```

Comment:
Might be good to support also other mechanisms. We try to make things generally usable even outside of Strimzi if the effort is reasonable.

Also, how will you get from `listener.name` to bootstrap address? Wouldn't it be better to have the address configured directly?

Reply:
Currently to construct the bootstrap address in our fork we call `InetAddress.getLocalHost().getCanonicalHostName()` to get the host (else fallback to 127.0.0.1 on error) and add the listener port to that.

To support other mechanisms we could allow the user to pass `client.quota.callback.kafka.config.*` properties and prefer these over our conveniences if present.

So, for example. We could default to using our internal bootstrap address construction code but prefer to use `client.quota.callback.kafka.config.bootstrap.servers` if it was set.

Reply:
I think we should just use user configured bootstrap server. Not sure `getCanonicalHostName` can be rellied on in general and how many issues around things such as TLS hostname verification will it arise from this. Making the ocnfiguration explicit and required will make this more transparent to the user.

Also, leaving that asside ... what do you need the `client.quota.callback.kafka.listener.name` for?

Reply:
We use the listener name to lookup the SSL configuration (truststore/keystore properties) from the broker configuration and then set them on the admin client configuration.

Reply:
How exactly do you plan to lookup the configuration? How do you know what to look for?

Reply:
Again talking about what we've done in our fork:

We take the configured listener name and then look up:
* `listener.name.${name}.ssl.keystore.location`
* `listener.name.${name}.ssl.keystore.password`
* `listener.name.${name}.ssl.truststore.location`
* `listener.name.${name}.ssl.truststore.password`
in the broker config and set them on the admin client config if they were present (after removing `listener.name.${name}`).

Using explicit configuration instead sounds good to me. 

Does strimzi-kafka-operator control all these for the replication listener? I.e. would it be able to set `bootstrap.servers`, `security.protocol` and the above keystore/truststore props?